File: dart/lib/src/bindings/create_reverse_name.dart
/// Create reverse name binding for SNS domains
///
/// This function creates reverse lookup entries for domains, mirroring
/// js/src/bindings/createReverseName.ts functionality exactly.
library;

import 'package:solana/solana.dart';
import '../constants/addresses.dart';
import '../instructions/create_reverse_instruction.dart';
import '../instructions/instruction_types.dart';
import '../utils/get_domain_key_sync.dart';

/// Parameters for creating a reverse name entry
class CreateReverseNameParams {
  const CreateReverseNameParams({
    required this.nameAccount,
    required this.name,
    required this.feePayer,
    this.parentName,
    this.parentNameOwner,
  });

  /// The name account to create the reverse account for
  final Ed25519HDPublicKey nameAccount;

  /// The name of the domain
  final String name;

  /// The fee payer of the transaction
  final Ed25519HDPublicKey feePayer;

  /// Optional parent name account
  final Ed25519HDPublicKey? parentName;

  /// Optional parent name owner
  final Ed25519HDPublicKey? parentNameOwner;
}

/// Creates reverse lookup entries for domains
///
/// This function mirrors js/src/bindings/createReverseName.ts exactly:
/// - Creates hashed reverse lookup from name account
/// - Derives reverse lookup account using CENTRAL_STATE
/// - Builds create reverse instruction with proper accounts
///
/// Returns list of transaction instructions for reverse name creation.
Future<List<TransactionInstruction>> createReverseName(
  CreateReverseNameParams params,
) async {
  // Create hashed reverse lookup from name account (matches JS logic)
  final hashedReverseLookup = getHashedNameSync(params.nameAccount.toBase58());

  // Derive reverse lookup account using centralState as parent
  final reverseLookupAccount = getNameAccountKeySync(
    hashedReverseLookup,
    nameClass: centralState,
    nameParent: params.parentName?.toBase58(),
  );

  // Create the reverse instruction parameters
  final instructionParams = CreateReverseInstructionParams(
    domain: params.name,
    programAddress: registryProgramAddress,
    namingServiceProgram: nameProgramAddress,
    rootDomain: rootDomainAddress,
    reverseLookup: await reverseLookupAccount,
    systemProgram: systemProgramAddress,
    centralState: centralState,
    payer: params.feePayer.toBase58(),
    rentSysvar: sysvarRentAddress,
    parentAddress: params.parentName?.toBase58(),
    parentOwner: params.parentNameOwner?.toBase58(),
  );

  // Create the reverse instruction
  final createReverseInstr = CreateReverseInstruction(
    domain: params.name,
    params: instructionParams,
  );

  // Build and return the instruction
  final instruction = createReverseInstr.build();

  return [instruction];
}

File: js/src/bindings/createReverseName.ts
import { PublicKey, SystemProgram, SYSVAR_RENT_PUBKEY } from "@solana/web3.js";
import { createReverseInstruction } from "../instructions/createReverseInstruction";
import {
  NAME_PROGRAM_ID,
  SNS_ROOT_DOMAIN_ACCOUNT,
  REGISTER_PROGRAM_ID,
  CENTRAL_STATE,
} from "../constants";
import { getHashedNameSync } from "../utils/getHashedNameSync";
import { getNameAccountKeySync } from "../utils/getNameAccountKeySync";

/**
 * Builds the instruction to create an SNS reverse lookup account for an
 * already-derived name account.
 *
 * This is a low-level SNS registrar helper: it creates reverse lookup accounts
 * for SNS names only. It is not suffix-aware and does not derive `nameAccount`
 * from a `.sns` domain string. The `name` argument is stored as provided and is
 * not validated, so callers must ensure it matches the supplied SNS
 * `nameAccount`. For subdomains, pass the parent name account and parent owner
 * so the reverse lookup is derived in the parent namespace.
 *
 * @param nameAccount The pre-derived SNS name account the reverse lookup points to
 * @param name The raw reverse name to store without a TLD suffix
 * @param feePayer The account paying for reverse account creation
 * @param parentName Optional parent name account, required for subdomain reverse lookups
 * @param parentNameOwner Optional parent name owner, required when `parentName` is provided
 * @returns An array containing the reverse lookup creation instruction
 */
export const createReverseName = async (
  nameAccount: PublicKey,
  name: string,
  feePayer: PublicKey,
  parentName?: PublicKey,
  parentNameOwner?: PublicKey,
) => {
  let hashedReverseLookup = getHashedNameSync(nameAccount.toBase58());
  let reverseLookupAccount = getNameAccountKeySync(
    hashedReverseLookup,
    CENTRAL_STATE,
    parentName,
  );

  let initCentralStateInstruction = new createReverseInstruction({
    name,
  }).getInstruction(
    REGISTER_PROGRAM_ID,
    NAME_PROGRAM_ID,
    SNS_ROOT_DOMAIN_ACCOUNT,
    reverseLookupAccount,
    SystemProgram.programId,
    CENTRAL_STATE,
    feePayer,
    SYSVAR_RENT_PUBKEY,
    parentName,
    parentNameOwner,
  );

  let instructions = [initCentralStateInstruction];

  return instructions;
};
